### PR TITLE
use custom mocha reporter to generate annotations

### PR DESCRIPTION
**Pull request by burmudar**
This extends the Mocha Spec reporter to be able to output the test summary to file so that we can generate an accurate annotation depicting the exact failures instead of crafting some complex regex.

This is how the annotation looks when it is generated
image

- [x] clean up test/debugging code
- [x] correct naming for all the things
- [x] correct location for all the things

## Test plan
Local testing and main-dry-running

### CODE REVIEW

**Review by valerybugakov — COMMENTED** [review]
Beautiful stuff @burmudar!

It would be awesome to leave a comment above the `SpecFileReporter` declaration with a high-level overview of what this reporter is doing under the hood and maybe link to a few useful resources for understanding the logic. I found a couple doing the review: 

1. [Tutorial: Create a Custom Reporter](https://mochajs.org/api/tutorial-custom-reporter.html)
2. [The Base reporter implementation](https://mochajs.org/api/reporters_base.js.html)

File: package.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
     "cover-integration": "nyc --hook-require=false --silent yarn test-integration",
     "cover-browser-integration": "nyc --hook-require=false --silent yarn workspace @sourcegraph/browser test-integration",
     "test-lighthouse": "SOURCEGRAPH_API_URL=https://sourcegraph.com lhci collect && lhci open",
-    "test-e2e": "TS_NODE_PROJECT=client/web/src/end-to-end/tsconfig.json mocha ./client/web/src/end-to-end/**/*.test.ts",
+    "test-e2e": "TS_NODE_PROJECT=client/web/src/end-to-end/tsconfig.json mocha --config .mocharc.js ./client/web/src/end-to-end/**/*.test.ts",
```
[the remark this answers is not in the document]
Comment (burmudar) [reply package.json:38]: While getting this to work I had some suspicion that the current working directory for some tests were different, but I'll remove it as test 😄

File: client/web/dev/reporter.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    if (this.buildkite) {
+      this.console = new Console({
+        stdout: fs.createWriteStream(`./annotations/mocha-test-output-${this.title}`),
+      })
+      let tmp = Base.consoleLog
+
+      let log = this.console.log
+      Base.consoleLog = log
+      super.epilogue()
+      Base.consoleLog = tmp
```
[the remark this answers is not in the document]
Comment (burmudar) [reply client/web/dev/reporter.js:35]: ooooh 👏🏼 this is awesome thanks!

File: .mocharc.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ module.exports = {
     __dirname + '/client/shared/dev/fetch',
     __dirname + '/client/shared/dev/suppressPollyErrors',
   ],
+  reporter: 'reporter.js',
```
[the remark this answers is not in the document]
Comment (burmudar) [reply .mocharc.js:8]: yeah I think that is definitely better! I'll move it 😄

File: client/web/dev/reporter.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    console.log('🚓 🚒 🧨 REPORTER')
+    this.title = 'placeholder'
+    this.buildkite = false
+
+    if ('BUILDKITE' in process.env) {
+      this.buildkite = true
+    }
+
+    if ('BUILDKITE_LABEL' in process.env) {
+      this.title = process.env.BUILDKIATE_LABEL
```
[the remark this answers is not in the document]
Comment (burmudar) [reply client/web/dev/reporter.js:18]: Good point. I'll log a warning, since everything should still work and this should not block any testing

**Review by valerybugakov — APPROVED** [review]
🎉 🔥

**Pull request merged**